larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,237 @@
+        data.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(x, densityDatum, accumulationDatum, deltaX));
+    }
+    const float x(static_cast<float>(inputData.back().first));
+    const float deltaX(m_xUpperBound-x);
+    const float densityDatum(static_cast<float>(inputData.back().second) / normalisation);
+    accumulationDatum+=densityDatum*(m_useWidths ? deltaX : 1.f);
+    data.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(x, densityDatum, accumulationDatum, deltaX));
+
+    return data;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::ResampleDiscreteProbabilityData(
+    const DiscreteProbabilityVector &discreteProbabilityVector, const ResamplingPoints &resamplingPoints) const
+{
+    if (2 > resamplingPoints.size())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    DiscreteProbabilityData resampledProbabilityData;
+
+    float prevCumulativeData(0.f);
+    for (size_t iSample = 0; iSample < resamplingPoints.size()-1; ++iSample)
+    {
+        const float xResampled(resamplingPoints.at(iSample));
+        const float deltaX(resamplingPoints.at(iSample+1)-xResampled);
+        const float cumulativeDatumResampled(discreteProbabilityVector.EvaluateCumulativeProbability(xResampled));
+        const float densityDatumResampled((cumulativeDatumResampled-prevCumulativeData)/(m_useWidths ? deltaX : 1.f));
+        resampledProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xResampled, 
+            densityDatumResampled, cumulativeDatumResampled, deltaX));
+        prevCumulativeData = cumulativeDatumResampled;
+    }
+
+    const float xResampled(resamplingPoints.back());
+    const float deltaX(m_xUpperBound-xResampled);
+    const float cumulativeDatumResampled(discreteProbabilityVector.EvaluateCumulativeProbability(xResampled));
+    const float densityDatumResampled((cumulativeDatumResampled-prevCumulativeData)/(m_useWidths ? deltaX : 1.f));
+    resampledProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xResampled, densityDatumResampled, 
+        cumulativeDatumResampled, deltaX));
+
+    return resampledProbabilityData;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::RandomiseDiscreteProbabilityData(
+   const DiscreteProbabilityVector &discreteProbabilityVector, std::mt19937 &randomNumberGenerator) const
+{
+    DiscreteProbabilityData randomisedProbabilityData;
+
+    std::vector<size_t> randomisedElements(discreteProbabilityVector.GetSize());
+    std::iota (std::begin(randomisedElements), std::end(randomisedElements), 0);
```

Comment:
Typically no space between function name and bracket.

Reply:
I love it.  I got it.